src/SourceCode.Clay.Collections/ListExtensions.cs

```diff
@@ -0,0 +1,160 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace SourceCode.Clay.Collections.Generic
+{
+    /// <summary>
+    /// Represents extensions for lists.
+    /// </summary>
+    public static class ListExtensions
+    {
+        /// <summary>
+        /// Performs an efficient item-by-item comparison.
+        /// </summary>
+        /// <typeparam name="T">The type of items.</typeparam>
+        /// <param name="x">List 1</param>
+        /// <param name="y">List 2</param>
+        /// <param name="comparer">The comparer to use to test for equality.</param>
+        /// <param name="colocated">Optimizes algorithm for cases when the inputs are expected to be ordered in the same manner.</param>
+        /// <returns></returns>
+        public static bool FastEquals<T>(this IReadOnlyList<T> x, IReadOnlyList<T> y, IEqualityComparer<T> comparer, bool colocated)
+        {
+            if (comparer == null) throw new ArgumentNullException(nameof(comparer));
+
+            // If one is null but not the other (xor), then not equal
+            if (ReferenceEquals(x, null) ^ ReferenceEquals(y, null)) return false;
```

Comment:
Benchmark this optimization. Null is usually rare and XOR works poorly with the branch predictor. Convert to this if it works out poorly:

```
if (ReferenceEquals(x, null) && ReferenceEquals(y, null)) return true;
if (ReferenceEquals(x, null) || ReferenceEquals(y, null)) return false;
```